MalformedURLException with Quarkus 1.12 (or master) when the app in `,;~!@#$%^&()` directory

Getting MalformedURLException with Quarkus 1.12 (or master) when the app in `,;~!@#$%^&()` directory. Works well with Quarkus 1.11.
Quarkus 1.12 - tests are working, dev mode is working, but running `java -jar ...` fails. Using JDK 11.
Seems `#` character is problematic for SRYE Config, `Caused by: java.net.MalformedURLException: no !/ found in url spec:file:/Users/rsvoboda/tmp/,;~!@` mentioned just part of `,;~!@#$%^&()` directory

Reproducer:
```
mkdir ',;~!@#$%^&()'
cd ',;~!@#$%^&()'
mvn io.quarkus:quarkus-maven-plugin:1.12.0.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.getting.started.GreetingResource" \
    -Dpath="/hello"
cd getting-started
mvn package
java -jar target/quarkus-app/quarkus-run.jar
```

Stacktrace:
```
$ java -jar target/quarkus-app/quarkus-run.jar
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:48)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:25)
Caused by: java.util.ServiceConfigurationError: org.eclipse.microprofile.config.spi.ConfigSource: Error accessing configuration file
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:581)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.parse(ServiceLoader.java:1172)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1205)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1220)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
	at io.smallrye.config.SmallRyeConfigBuilder.discoverSources(SmallRyeConfigBuilder.java:90)
	at io.smallrye.config.SmallRyeConfig.buildConfigSources(SmallRyeConfig.java:83)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:66)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:358)
	at io.quarkus.runtime.generated.Config.<clinit>(Config.zig:326)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:57)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	... 6 more
Caused by: java.net.MalformedURLException: no !/ found in url spec:file:/Users/rsvoboda/tmp/,;~!@
	at java.base/java.net.JarURLConnection.parseSpecs(JarURLConnection.java:172)
	at java.base/java.net.JarURLConnection.<init>(JarURLConnection.java:158)
	at java.base/sun.net.www.protocol.jar.JarURLConnection.<init>(JarURLConnection.java:81)
	at java.base/sun.net.www.protocol.jar.Handler.openConnection(Handler.java:41)
	at java.base/java.net.URL.openConnection(URL.java:1074)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.parse(ServiceLoader.java:1162)
	... 26 more
```

--- a/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java
+++ b/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java
@@ -108,9 +108,14 @@ public class JarResource implements ClassLoadingResource {
                 if (realName.endsWith("/")) {
                     realName = realName.substring(0, realName.length() - 1);
                 }
-                URI jarUri = jarPath.toUri();
-                return new URL("jar", null, jarUri.getScheme() + ':' + jarUri.getPath() + "!/" + resource);
-            } catch (MalformedURLException e) {
+                final URI jarUri = jarPath.toUri();
+                // first create a URI which includes both the jar file path and the relative resource name
+                // and then invoke a toURL on it. The URI reconstruction allows for any encoding to be done
+                // for the "path" which includes the "realName"
+                final URL resUrl = new URI(jarUri.getScheme(), jarUri.getPath() + "!/" + realName, null).toURL();
+                // wrap it up into a "jar" protocol URL
+                return new URL("jar", null, resUrl.getProtocol() + ':' + resUrl.getPath());
+            } catch (MalformedURLException | URISyntaxException e) {
                 throw new RuntimeException(e);
             }
         } finally {

--- a/independent-projects/bootstrap/runner/src/test/java/io/quarkus/bootstrap/runner/JarResourceTest.java
+++ b/independent-projects/bootstrap/runner/src/test/java/io/quarkus/bootstrap/runner/JarResourceTest.java
@@ -1,0 +1,57 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.exporter.ZipExporter;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests {@link JarResource}
+ */
+public class JarResourceTest {
+
+    /**
+     * Tests that the URL(s) returned from {@link JarResource#getResourceURL(String)} are properly encoded and can be used
+     * to open connection to the URL to read data
+     */
+    @Test
+    public void testResourceURLEncoding() throws Exception {
+        final JavaArchive jar = ShrinkWrap.create(JavaArchive.class);
+        final String[] files = new String[] { "a.txt", "a b.txt", ",;~!@#$%^&().txt" };
+        for (final String file : files) {
+            jar.add(new StringAsset("hello"), file);
+        }
+        final Path testDir = Files.createTempDirectory("test");
+        // create a child dir with special characters
+        final Path specialCharDir = Files.createDirectory(Paths.get(testDir.toString(), ",;~!@#$%^&()"));
+        final Path jarFilePath = Files.createTempFile(specialCharDir, "test", "quarkus-test.jar");
+        // create a jar file under the directory which has the special characters
+        jar.as(ZipExporter.class).exportTo(jarFilePath.toFile(), true);
+        final JarResource jarResource = new JarResource(null, jarFilePath);
+        for (final String resource : files) {
+            final URL url = jarResource.getResourceURL(resource);
+            Assertions.assertNotNull(url, resource + " is missing in jar");
+            // check that opening the resource URL works and data can be read
+            final URLConnection conn = url.openConnection();
+            try (final InputStream is = conn.getInputStream()) {
+                drainFully(is);
+            }
+        }
+    }
+
+    private static void drainFully(final InputStream inputStream) throws IOException {
+        int read = -1;
+        final byte[] data = new byte[1024];
+        while ((read = inputStream.read(data)) != -1) {
+        }
+    }
+}

